test(plugin-webpack): minimal test

gr2m opened the pull request:
This is part of https://github.com/pikapkg/snowpack/issues/1019

## Changes

Adding tests to [`plugins/plugin-webpack`](https://github.com/pikapkg/snowpack/tree/master/plugins/plugin-webpack)

gr2m commented:
this is not as straight forward as the tests for [the plugin-vue tests](https://github.com/pikapkg/snowpack/tree/d36ddd82d95b61632ebdabe6a3876e9a3df37fd7/plugins/plugin-vue/test), because Webpack's optimize plugin is actually mutating file. I've mocked `fs.writeFile` and `fs.writeFileSync` in a minimal test:

https://github.com/pikapkg/snowpack/blob/eedb38dde3d367ed005b141d684d31cf945c465d/plugins/plugin-webpack/test/plugin.test.js#L6-L27

Does that approach look good to you @drwpow @FredKSchott?

drwpow commented on plugins/plugin-webpack/plugin.js:
```diff
@@ -375,8 +363,8 @@ module.exports = function plugin(config, args = {}) {
-                }
-          )
+                },
+          ),
```
Bah Prettier changes… thanks for updating this

gr2m replied:
I'm sorry for all the unrelated changes :(

drwpow commented on plugins/plugin-webpack/test/plugin.test.js:
```diff
@@ -0,0 +1,31 @@
+    buildDirectory: path.resolve(__dirname, 'stubs/minimal/'),
+  });
+
+  expect(fs.writeFileSync.mock.calls.map(toPathAndStringContent)).toMatchSnapshot(
```
Nice! I like this approach. The danger with these snapshot tests (sometimes) is that if a file failed to generate, the test would pass because it would only test snapshots for what was generated. But here we’re doing a test specifically to ensure all files are generated before comparing contents, which is perfect.

drwpow commented on plugins/plugin-webpack/test/plugin.test.js:
```diff
@@ -0,0 +1,31 @@
+
+const plugin = require('../plugin');
+
+test('minimal - no options', async () => {
```
Not a request for change, but Jest makes it helpful to group tests with `describe()` for the group, and `it()` for the individual test. But here, this is fine—no grouping necessary.

drwpow replied:
Also I like the test description here: `minimal - no options`. I’m +1 in general for very short, to-the-point test names which are much easier to read than a long, long list of complete sentences with complex structure.

drwpow reviewed (approved):
This LGTM! Thanks for adding this!

ralphtheninja commented on plugins/plugin-webpack/README.md:
```diff
@@ -14,16 +14,30 @@ Add `@snowpack/plugin-webpack` to `snowpack.config.json`:
 module.exports = {
-  plugins: [["@snowpack/plugin-webpack", { /* see "Plugin Options" below */ }]]
-}
+  plugins: [
```
Yay. Better docs on top of my changes! :smile:

gr2m replied:
oh my I'm sorry for the unrelated changes. I've `prettier` setup to format on save, and didn't pay attention.

ralphtheninja replied:
No no, I think it's fine :)

gr2m committed:
test(plugin-webpack): minimal test

gr2m committed:
test(@snowpack/plugin-webpack): use describe/it

resolves https://github.com/pikapkg/snowpack/pull/1061\#discussion_r489707178 /cc @drwpow

gr2m commented:
I'll merge this in, then add some more tests for the different options in a follow up PR

gr2m commented:
The CI gets stuck and the tests fail locally now with some weird error 

> ENOENT: no such file or directory, rename '/private/var/folders/hs/x9qtfmvn1lz1sgml9q21h7k80000gn/T/tmp.GNGoo3JE/snowpack/node_modules/.cache/terser-webpack-plugin/tmp/67975a4e' -> '/private/var/folders/hs/x9qtfmvn1lz1sgml9q21h7k80000gn/T/tmp.GNGoo3JE/snowpack/node_modules/.cache/terser-webpack-plugin/content-v2/sha512/7e/76/a9131f88825b28596d0158e0f7bcdea7fe5a3ac941d2f5dbccc3c5504499a6bd443643d19b4324a9e0fd7a9d5be1a68d3e58ee61ccf828d3ccbeedabf2f8'

I've rebased on latest master, it didn't do that before. I'm investigating ...

MoonBall commented:
I run `yarn test -t @snowpack/plugin-webpack` on my machine, and the program throw the following error. I don't know why.
```
 FAIL  plugins/plugin-webpack/test/plugin.test.js
  ● @snowpack/plugin-webpack › minimal - no options

    ENOENT: no such file or directory, rename '/Users/chengang.07/git-source/github.com/pikapkg/snowpack/node_modules/.cache/terser-webpack-plugin/tmp/3aba3b6a' -> '/Users/chengang.07/git-source/github.com/pikapkg/snowpack/node_modules/.cache/terser-webpack-plugin/content-v2/sha512/7e/76/a9131f88825b28596d0158e0f7bcdea7fe5a3ac941d2f5dbccc3c5504499a6bd443643d19b4324a9e0fd7a9d5be1a68d3e58ee61ccf828d3ccbeedabf2f8'
```

gr2m commented:
Thanks for checking @MoonBall, I'm digging into this today

gr2m committed:
workaround "no such file or directory" error and hanging CI

gr2m commented:
I had to add a few workarounds. The `fs.{writeFile,writeFileSync}` calls were different depending on whether cache was written or not. I changed the code to pass through all operations that are not within the current folder, and filter out these calls from the snapshot. 

@drwpow Would you mind checking again if it all makes sense?

gr2m committed:
unix-ify folder separators for Windows

gr2m committed:
test(plugin-webpack): all options

gr2m commented:
I've added another test for all options. 

Questions

- Do you think we need tests for `buildOptions`?
- We currently test all files written in the `/plugins/plugin-webpack/test/` folder, including files such as `webpack-runtime.js`, which feels like testing too much. I'd say it's a problem when it's a problem, but if you want I can setup a ignore list so we don't do snapshots for these calls
- Any thoughts on coverage in the `plugins/` folder? I'm a big fan of 100% test coverage, I think we could achieve that for the plugins, it would allow us to accept plugin contributions with more confidence. I _think_ jest allows to enable coverage in only a few selected folders

@drwpow @FredKSchott  @melissamcewen any thoughts ☝🏼

gr2m committed:
test(plugin-webpack): unixify newlines

gr2m commented:
The build currently fails because `index-0.js.map` is 188 bytes on UNIX, but 190 bytes on Windows 🤦🏼 Most likely because of the `\\r\\n` while UNIX has `\\n`. 

Any thought on how to easily resolve this, other than normalizing byte sizes using regex in the `console.log` calls snapshots?

drwpow commented:
> The build currently fails because `index-0.js.map` is 188 bytes on UNIX, but 190 bytes on Windows 🤦🏼 Most likely because of the `\\r\\n` while UNIX has `\\n`.

This is a semi-similar problem to #1048: it’s hard to resolve tests for different platforms well, especially when we want to check contents.

Unfortunately I don’t know of a better way to solve this other than trying to coerce all the Windows differences to conform to Linux (like all the helpers we have in `tests/build.test.js`)

gr2m commented:
> like all the helpers we have in `tests/build.test.js`

you mean this?

https://github.com/pikapkg/snowpack/blob/590cbfdb4b4029131353dcae9b11ae1bceb5364b/test/build/build.test.js#L12-L17

@FredKSchott you mentioned you had normalization for byte numbers in output already? I couldn't find that yet

gr2m commented:
I think we could create a custom serializer that encapsulates all the normalization logic with need for snowpack, similar to https://github.com/tribou/jest-serializer-path

gr2m committed:
test(plugin-webpack): normalize console.log output

gr2m committed:
test(plugin-webpack): use custom jest snapshot serializer

gr2m committed:
test: move jest snapshot serializer into separate file for future reusablity

gr2m committed:
test: simplify snapshot names

gr2m committed:
test(plugin-webpack): sort files in snapshot string by path

gr2m committed:
test(plugin-webpack): ignore writes outside the test folder in snapshot

gr2m commented:
I think I'm all set for a first set of tests. [I created a Jest snapshot serializer](https://github.com/pikapkg/snowpack/blob/1019/plugin-webpack-tests/plugins/plugin-webpack/test/serializer.js) which we might be able to reuse in other tests. The snapshots are much more concise as a result: https://github.com/pikapkg/snowpack/blob/1019/plugin-webpack-tests/plugins/plugin-webpack/test/__snapshots__/plugin.test.js.snap.

I'll have a look at the other tests to see whether it makes sense to use re-use the serializer, or maybe at least parts of it such as file content normalization.

drwpow commented on plugins/plugin-webpack/test/serializer.js:
```diff
@@ -0,0 +1,71 @@
+function toPathAndStringContent(basePath, [path, content]) {
+  const shortPath = path.replace(basePath, '').substr(1);
+  // unix-ify folder separators for Windows
+  const normalizedPath = shortPath.replace(/\\/g, '/');
```
Could this possibly be replacing a backslash out-of-context when it doesn’t need to? I think it’s probably fine; just wanted to raise that.

gr2m replied:
I don't think so, the replace is only run on the `path` argument of `fs.writeFile`/`fs.writeFileSync`, not the content or log output

drwpow replied:
Ah OK. Yeah I ask because we had a problem with Rollup where it inlined comments with things like `/* src\App.js */` in the actual code, so we had to find a way to replace those on Windows (I think we just ended up stripping out comments for comparison). Moral of the story is you never know where backslashes will pop up! 🙃

drwpow reviewed (approved):
Awesome! This looks great. Yeah I really like the idea of the snapshot serializer here, and think we could definitely use it to clean up some of our Snowpack snapshots.

We do have to conditionally apply/skip some replacement functions in different parts, but I see you’ve covered that here with `if (value.getMockName() === …`. We might have to do something slightly-different, but same basic idea. Anyway, would love to get this merged!
